ZOOKEEPER-4235 Java Client SendThread does not clean up created objects during constructor of SaslClient and Login

rvaleti opened the pull request:
Change Description - Move initialization of ZookeeperSaslClient & Login to instance methods.

- Moved ZookeeperSaslClient's initialization from constructor to  instance method createSaslClient()
- Moved Login's initialization from constructor to an instance method loginAndStartThreadIfRequired()
- Moved the anonymous Runnable implementation of login thread to LoginRunnable inner class.

dbwong commented:
Overall looks okay to me.  Do you think it is worth making this code more testable via introduction of a factory pattern (or alternatively something like power mock)?  And verifying the shutdown paths now work/didn't work?

eolivelli reviewed (commented):
I have to take a look more closely, but an alternative way would be to catch the error in the constructor and perform clean up before failing the constructor.

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -1148,7 +1148,8 @@ private void startConnect(InetSocketAddress addr) throws IOException {
                         zooKeeperSaslClient.shutdown();
                     }
-                    zooKeeperSaslClient = new ZooKeeperSaslClient(SaslServerPrincipal.getServerPrincipal(addr, clientConfig), clientConfig);
+                    zooKeeperSaslClient = new ZooKeeperSaslClient(clientConfig);
+                    zooKeeperSaslClient.createSaslClient(SaslServerPrincipal.getServerPrincipal(addr, clientConfig));
```
What's the point of this change?
You split the logic into 2 methods which you call one after the other. I don't see how this makes the code more resource efficient.

eolivelli replied:
The benefit is that you will never see the constructor fail and then you can have a handle to perform cleanup

anmolnar replied:
Got it.

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/Login.java:
```diff
@@ -95,10 +95,159 @@
+
+    public void loginAndStartThreadIfRequired() throws LoginException {
         login = login(loginContextName);
-        this.loginContextName = loginContextName;
         subject = login.getSubject();
         isKrbTicket = !subject.getPrivateCredentials(KerberosTicket.class).isEmpty();
         AppConfigurationEntry[] entries = Configuration.getConfiguration().getAppConfigurationEntry(loginContextName);
```
You moved this logic from the constructor to the new method `loginAndStartThreadIfRequired()` which could be beneficial if the constuction of the Login object and the method call would be separated, but I don't see that would be the case.

Login class creation always goes together with startIfNeeded(), so it doesn't really matter what happens in which. I'd like to see what exactly the resource which is not cleared properly. The thread object?

rvaleti replied:
Yes, the refresh thread of Login object is not cleaned up leaving many login threads.
The Jira's description is missing one key stack trace

```
17971 Threads with the following stack trace

3911 - TIMED_WAITING - Thread-822 | ORG: null RAC: None
java.base@11.0.9/java.lang.Thread.sleep(Native Method)
org.apache.zookeeper.Login$1.run(Login.java:197)
java.base@11.0.9/java.lang.Thread.run(Thread.java:834)
```

which when combined with another stack trace coming from ZookeeperSaslClient

```
java.lang.Thread.State: RUNNABLE at java.net.PlainSocketImpl.socketConnect(java.base@11.0.4.0.101/Native Method) at java.net.AbstractPlainSocketImpl.doConnect(java.base@11.0.4.0.101/AbstractPlainSocketImpl.java:399) - locked <0x00000015004fde20> (a java.net.SocksSocketImpl) at java.net.AbstractPlainSocketImpl.connectToAddress(java.base@11.0.4.0.101/AbstractPlainSocketImpl.java:242) at java.net.AbstractPlainSocketImpl.connect(java.base@11.0.4.0.101/AbstractPlainSocketImpl.java:224) at java.net.SocksSocketImpl.connect(java.base@11.0.4.0.101/SocksSocketImpl.java:403) at java.net.Socket.connect(java.base@11.0.4.0.101/Socket.java:609) at sun.security.krb5.internal.TCPClient.<init>(java.security.jgss@11.0.4.0.101/NetClient.java:62) at sun.security.krb5.internal.NetClient.getInstance(java.security.jgss@11.0.4.0.101/NetClient.java:42) at sun.security.krb5.KdcComm$KdcCommunication.run(java.security.jgss@11.0.4.0.101/KdcComm.java:401) at sun.security.krb5.KdcComm$KdcCommunication.run(java.security.jgss@11.0.4.0.101/KdcComm.java:364) at java.security.AccessController.doPrivileged(java.base@11.0.4.0.101/Native Method) at sun.security.krb5.KdcComm.send(java.security.jgss@11.0.4.0.101/KdcComm.java:348) at sun.security.krb5.KdcComm.sendIfPossible(java.security.jgss@11.0.4.0.101/KdcComm.java:253) at sun.security.krb5.KdcComm.send(java.security.jgss@11.0.4.0.101/KdcComm.java:234) at sun.security.krb5.KdcComm.send(java.security.jgss@11.0.4.0.101/KdcComm.java:200) at sun.security.krb5.KrbAsReqBuilder.send(java.security.jgss@11.0.4.0.101/KrbAsReqBuilder.java:326) at sun.security.krb5.KrbAsReqBuilder.action(java.security.jgss@11.0.4.0.101/KrbAsReqBuilder.java:371) at com.sun.security.auth.module.Krb5LoginModule.attemptAuthentication(jdk.security.auth@11.0.4.0.101/Krb5LoginModule.java:754) at com.sun.security.auth.module.Krb5LoginModule.login(jdk.security.auth@11.0.4.0.101/Krb5LoginModule.java:592) at javax.security.auth.login.LoginContext.invoke(java.base@11.0.4.0.101/LoginContext.java:726) at javax.security.auth.login.LoginContext$4.run(java.base@11.0.4.0.101/LoginContext.java:665) at javax.security.auth.login.LoginContext$4.run(java.base@11.0.4.0.101/LoginContext.java:663) at java.security.AccessController.doPrivileged(java.base@11.0.4.0.101/Native Method) at javax.security.auth.login.LoginContext.invokePriv(java.base@11.0.4.0.101/LoginContext.javaa:663) at javax.security.auth.login.LoginContext.login(java.base@11.0.4.0.101/LoginContext.java:574) at org.apache.zookeeper.Login.login(Login.java:304) - locked <0x000000151c477148> (a org.apache.zookeeper.Login) at org.apache.zookeeper.Login.<init>(Login.java:106) at org.apache.zookeeper.client.ZooKeeperSaslClient.createSaslClient(ZooKeeperSaslClient.java:249) - locked <0x000000151c476f68> (a org.apache.zookeeper.client.ZooKeeperSaslClient) at org.apache.zookeeper.client.ZooKeeperSaslClient.<init>(ZooKeeperSaslClient.java:141) at org.apache.zookeeper.ClientCnxn$SendThread.startConnect(ClientCnxn.java:972) at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1031)
```

Here is one possibility of Login threads may not cleared by shutdown()

1. Zookeeper initialization creates & starts ClientCnxn -> spins up SendThread and starts it
2. ClientCnxn.SendThread.run() calls ZookeeperSaslClient.<init> -> createSaslClient()
3. Login.<init> -> Login.login() -> loginContext.login()
4. Let's say thread1 executing a login (loginContext.login()) takes few seconds.
5. Meantime, let's say some one called Zookeeper.close() -> ClientCnxn.close() -> ClientCnxn.disconnect()
6. At this point, since ZooKeeperSaslClient's constructor is not yet finished, it's zooKeeperSaslClient.shutdown() is never called.
7. Once 3 get successful, a login thread refresh is created which would become a zombie.

So, the idea is 2 parts
1. Move the actual "login" which can take time in to separate methods and make ZooKeeperSaslClient and Login constructors lightweight so that we have a reference to these at all times.
2. Make sure either Login thread is interrupted (when shutdown called after the thread was started) OR not started (when shutdown called before the thread was started)

rvaleti replied:
@anmolnar , could you please review the updated files with above context?

anmolnar replied:
Doing another round of review, sorry for the delay.

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/Login.java:
```diff
@@ -125,150 +274,9 @@ public Login(final String loginContextName, CallbackHandler callbackHandler, fin
+        t.start();
     }
 
     public void startThreadIfNeeded() {
```
Do we still need this method?

eolivelli replied:
it looks like it is no more used

anmolnar reviewed (changes requested):
Please elaborate more on your change. What's the resource (which object or objects) that are not getting cleaned properly and how your patch solves that problem.

rvaleti commented:
> I have to take a look more closely, but an alternative way would be to catch the error in the constructor and perform clean up before failing the constructor.

@eolivelli , Please have a look at my other comment on the pretext of the issue. If looks good, do you mind taking a look at the review?

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/client/ZooKeeperSaslClient.java:
```diff
@@ -112,14 +112,17 @@ public String getLoginContext() {
+        this.clientConfig = clientConfig;
+    }
+
+    public void createSaslClient(final String serverPrincipal) throws LoginException {
```
We could keep the serverPrincipal into a field, this way we preserve the original behaviour, that this value is set during the construction.
Otherwise this method may be called with the wrong value.
Does it make sense to you?

rvaleti replied:
@eolivelli , Makes sense. Update the files in the review.

eolivelli reviewed (commented):
Now the patch looks good.
can you please answer to the question about removing `startThreadIfNeeded` ?

rvaleti commented:
> Now the patch looks good.
> can you please answer to the question about removing `startThreadIfNeeded` ?


My bad, I wanted to have it as a separate method so that we can take care of not starting if already marked for cancellation. But missed calling it earlier.

I updated the file with below changes
- renamed the startThreadIfNeeded to startLoginThread
- loginAndStartThreadIfRequired now calls startLoginThread to spin up the thread if not marked for cancellation already.

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/Login.java:
```diff
@@ -120,172 +272,36 @@ public Login(final String loginContextName, CallbackHandler callbackHandler, fin
-        if (t != null) {
+    private void startLoginThread() {
+        // Start the thread only if not canceled before reaching here.
+        synchronized (loginThreadLock) {
```
I think in Java you can use synchronized methods if the block encapsulates the entire method. You don't need to introduce a separate lock object and there's no performance hit either. I believe the syntax sugar uses the same locking method, but on the object itself.

rvaleti replied:
@anmolnar 
Yes, I used synchronized block initially, but realized there are already a bunch of synchronized methods in Login (for ex reLogin() which does a krb relogin). We should be able to cancel/interrupt the thread even if it's in the process of relogin. So, that's why I turned to a synchronized block.

anmolnar replied:
Got you.

anmolnar reviewed (approved):
lgtm. Thanks for the clarification.

anmolnar commented:
@rvaleti I restarted the Jenkins build for you.

anmolnar commented:
@rvaleti You got a green build now. Let's wait for @eolivelli 's approval and you're good to go.

rvaleti commented:
@eolivelli If you are happy with the PR, can you please approve?

eolivelli reviewed (approved):
LGTM

sorry for late reply

eolivelli commented:
I have forced a new run of the CI job

rvaleti commented:
@eolivelli , looks like previous build failure was due to server not coming up and not related to this change. And @anmolnar  forced a build earlier that succeeded. Do you want to re trigger the build or merge the PR?

rvaleti commented:
@eolivelli @anmolnar , could you please take a look at this & merge. We would like to cherrypick this in to our light fork at the earliest.

eolivelli commented:
Sorry for late reply.
I have restarted CI.
Committing as soon as the test pass.

thank you very much for this useful contribution !

mnpoonia commented:
@eolivelli Can you please help here. Is there something more to be done to get this PR through?

rvaleti commented:
> @rvaleti - Is this done. I was wondering if we can pick it in our light fork?

No pending work from the PR perspective. It's just that we need to get a green build (we had a green build once during the changes).

@eolivelli @anmolnar could you please re-trigger the build and see if it passes.

kezhuw commented:
Thank all for your work!

I am going to close this as #1755 should have fixed this.
